File: src/main/java/ca/mcgill/science/tepid/server/rest/JobDataMonitor.java
package ca.mcgill.science.tepid.server.rest;

import ca.mcgill.science.tepid.models.data.PrintJob;
import ca.mcgill.science.tepid.models.data.ViewResultMap;
import ca.mcgill.science.tepid.server.util.WebTargetsKt;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.ws.rs.client.Entity;
import javax.ws.rs.client.WebTarget;
import javax.ws.rs.core.MediaType;
import java.io.File;
import java.util.List;

public class JobDataMonitor implements Runnable {

    private final WebTarget couchdb = WebTargetsKt.getCouchdbOld();
    private static final Logger logger = LoggerFactory.getLogger(JobDataMonitor.class);

    private static class JobResultSet extends ViewResultMap<String, PrintJob> {
    }

    @Override
    public void run() {
        logger.info("Deleting expired job data.");
        try {
            List<PrintJob> jobs = couchdb.path("_design/main/_view").path("storedJobs")
                    .request(MediaType.APPLICATION_JSON).get(JobResultSet.class).getValues();
            jobs.forEach(j -> {
                if (j.getDeleteDataOn() < System.currentTimeMillis()) {
                    if (j.getFile() != null) {
                        try {
                            File f = new File(j.getFile());
                            if (f.exists())
                                if (!f.delete())
                                    System.out.println("Failed to delete file");
                        } catch (Exception ignored) {
                        }
                        j.setFile(null);
                    }
                    if (j.getAdditionalProperties().containsKey("_attachments")) {
                        j.getAdditionalProperties().remove("_attachments");
                    }
                    // todo verify; originally request(MediaType.TEXT_PLAIN)
                    couchdb.path(j.getId()).request(MediaType.TEXT_PLAIN).put(Entity.entity(j, MediaType.APPLICATION_JSON));
                }
            });
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

}

File: src/main/java/ca/mcgill/science/tepid/server/rest/JobDataMonitor.kt
package ca.mcgill.science.tepid.server.rest

import ca.mcgill.science.tepid.models.data.PrintJob
import ca.mcgill.science.tepid.server.util.CouchDb
import ca.mcgill.science.tepid.server.util.putJson
import ca.mcgill.science.tepid.utils.WithLogging
import java.io.File

class JobDataMonitor : Runnable {

    init {
        println("Init JobDataMonitor KT")
    }

    override fun run() {
        log.info("Deleting expired job data.")
        val now = System.currentTimeMillis()
        try {
            CouchDb.getViewRows<PrintJob>("storedJobs").forEach { j ->
                if (j.deleteDataOn < System.currentTimeMillis()) {
                    val filePath = j.file
                    if (filePath != null) {
                        try {
                            val f = File(filePath)
                            if (f.exists() && !f.delete())
                                log.error("Failed to delete file")
                        } catch (ignored: Exception) {
                        }

                        j.file = null
                    }

                    j.additionalProperties.remove("_attachments")
                    CouchDb.path(j._id).putJson(j)
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

        log.info("Deleting expired jobs completed in ${System.currentTimeMillis() - now} millis")
    }

    companion object : WithLogging()

}
